UI Support for global SetViews

fde31 opened the pull request:
This adds UI support for global, user defined parameters views in a set. A few things we might want to discuss probably:

### Naming

Currently everything uses `SetView` in order to refer to this new paradigm. Not sure that's what we to use in the end, it seems a bit too code-y, especially given the camelCase?

### SetView Management

I've moved the SetView management to a drawer, basically following the UX of presets and GraphSets. I think one aspect that is still missing from the UI currently is implementing `sort_order`.

<img width="784" alt="Screenshot 2024-12-20 at 19 13 03" src="https://github.com/user-attachments/assets/e2aa0d12-7f6f-4e68-9c65-4d29ed0d4dec" />

### Parameter Ordering / Management

I've moved param ordering and management into a dropdown and a modal for now. Would be great to get DnD in the list view in place (and maybe in other places of the app too) but thought we might not want to hold back the feature bc of that and could instead ticket as a follow up FR?

Also the modal UI is relatively primitive but if we want to change that or if it requires a some redesigning it shouldn't be too hard. Tbh would prefer a diff set of eyes or some user feedback here to determine what would be helpful.

<img width="619" alt="Screenshot 2024-12-20 at 19 13 36" src="https://github.com/user-attachments/assets/7ae3bc22-4cb7-4213-a67f-7aef99384291" />

![image](https://github.com/user-attachments/assets/35ccb240-885f-4a2d-a380-2a43478125af)

### View support for sets

Currently it's a bit confusing that one can be in a state that is considered "No Set" by the runner which also means that there is no support for SetViews. Maybe related to #182 but I wonder what your thoughts are about this and whether it should actually be possible to be in a "no set state" moving forward?

Setting to draft for now.

See #179

x37v commented on src/lib/util.ts:
```diff
@@ -1,7 +1,9 @@
 import { AnyJson, JsonMap, MIDIChannelPressureMetaMapping, MIDIControlChangeMetaMapping, MIDIKeypressMetaMapping, MIDIMetaMapping, MIDINoteMetaMapping, MIDIPitchBendMetaMapping, MIDIProgramChangeMetaMapping, OSCQueryStringValueRange, OSCQueryValueRange } from "./types";
 import { MIDIMetaMappingType } from "./constants";
 
+export const genericMemo: <P>(component: P) => P = memo;
+
```
unused?

fde31 replied:
nah, it's used here in order to allow wrapping a comp with generic props:
https://github.com/Cycling74/rnbo-runner-panel/pull/192/files#diff-a1d8d95fad145f23aabf4e0f183824c6d283093b806284f792f02458286ee80eR18

x37v commented on src/actions/sets.ts:
```diff
@@ -255,8 +296,360 @@ export const renameSetPresetOnRemote = (preset: PresetRecord, newname: string):
+				address: "/rnbo/inst/control/sets/views/create",
+				args: [
+					{ type: "s", value: name },
+					...params.valueSeq().map(p => ({ type: "s", value: p.setViewId })).toArray()
```
this seems to add parameters sort of randomly, i think they should be sorted by `instance_index` then `index` just like you get when you do the `Include All` in the set view param editor

x37v reviewed (commented):
looking great. A couple of little notes.
really happy that we can see if set view params are MIDI mapped, I'm sure folks are going to want to MIDI map from there so maybe we should create a ticket for that?
Seems like at least metadata editing should be easy to add right there?

fde31 commented:
MetaData editing isn't a problem yeah. If the runner can handle the fact that for a SetView we'd basically arm multiple instances at the same time for MIDI mapping via `/midi/last/report` then I might as well just add that right away?

x37v commented:
> MetaData editing isn't a problem yeah. If the runner can handle the fact that for a SetView we'd basically arm multiple instances at the same time for MIDI mapping via `/midi/last/report` then I might as well just add that right away?

I'm not 100% on that but.. might work?

fde31 commented:
OK addressed the comments above. Would still like to discuss the points from the initial comment:

* Managing `sort_order`
* Ensuring there is always a`Set` loaded
* naming for `SetViews`

Apart from that I've enabled the MIDI mapping features also on the SetView list. The only item I'd love to clean up is that we have this concept of `PatcherInstance.index` and `PatcherInstance.id` which are kinda obsolete / confusing and we might want to just consider the `index` to be the `id` similar to how we deal with `SetViews`

fde31 commented:
@x37v I've updated the PR to deal with the way `order` is now a distinct value on the OSCQuery state and additionally simplified the UX a bit by automatically loading a SetView upon creation if none was selected before.

x37v reviewed (approved):
looks great
